daemon/snap.go

```diff
@@ -338,12 +342,68 @@ func fillComponentInfo(about aboutSnap) []client.Component {
 	return comps
 }
 
-// snapIcon tries to find the icon inside the snap
+// snapIcon tries to find the icon inside the snap.
 func snapIcon(info snap.PlaceInfo) string {
 	found, _ := filepath.Glob(filepath.Join(info.MountDir(), "meta", "gui", "icon.*"))
-	if len(found) == 0 {
+	if len(found) > 0 {
+		return found[0]
+	}
+
+	// Try to find the icon via the Icon= field of a .desktop file for the snap.
+
+	// Find all .desktop files in ${SNAP}/meta/gui
+	desktopFiles, _ := filepath.Glob(filepath.Join(info.MountDir(), "meta", "gui", "*.desktop"))
+	if len(desktopFiles) == 0 {
 		return ""
 	}
 
-	return found[0]
+	// Try <snap>.desktop first
+	if snapDesktop := info.SnapName() + ".desktop"; strutil.ListContains(desktopFiles, snapDesktop) {
+		if iconPath, err := iconPathFromDesktopFile(info, snapDesktop); err == nil {
+			return iconPath
+		}
+	}
+
+	// Try <snap>_<snap>.desktop next
+	if snapAppDesktop := fmt.Sprintf("%s_%s.desktop", info.SnapName(), info.SnapName()); strutil.ListContains(desktopFiles, snapAppDesktop) {
+		if iconPath, err := iconPathFromDesktopFile(info, snapAppDesktop); err == nil {
+			return iconPath
```

Comment:
In my opinion, these two checks should be done before taking the first icon found at `meta/gui/*.icon`, because it is more probable that they have more meaning.

Reply:
They have more meaning for the user if the user is looking for a snap-app icon, but the purpose of the `/v2/icons/{name}/icon` endpoint is to return the icon of the snap itself, not any particular app.

Putting these checks later was my attempt to minimally change things: use the snap icon if it exists, fall back to a best guess otherwise. But I agree it would be nice to have something which looks up a snap-app icon directly via the .desktop file here.